### PR TITLE
Reuse 32 and 64-byte backing stores

### PR DESCRIPTION
**Pull request by BradleyWood**
Signed-off-by: BradleyWood <bradley.wood@ibm.com>

### CODE REVIEW

**Comment by BradleyWood** [comment]
Jenkins build all

**Comment by BradleyWood** [comment]
Jenkins build all

File: compiler/codegen/CodeGenRA.cpp [hunk 1]
```diff
@@ -458,24 +458,36 @@ OMR::CodeGenerator::allocateSpill(int32_t dataSize, bool containsCollectedRefere
...
-       spill = self()->getSpill16FreeList().front();
-       self()->getSpill16FreeList().pop_front();
-     }
+      {
+      if (dataSize <= 4 && !self()->getSpill4FreeList().empty())
+         {
+         spill = self()->getSpill4FreeList().front();
+         self()->getSpill4FreeList().pop_front();
+         }
+      if (!spill && try8ByteSpills && !self()->getSpill8FreeList().empty())
```
Comment (0xdaryl) [inline compiler/codegen/CodeGenRA.cpp:471]: Shouldn't all these `if`s be `else if`s instead?  I think only one of these cases would ever apply at any given moment.

File: compiler/codegen/CodeGenRA.cpp [hunk 2]
```diff
@@ -556,9 +568,9 @@ OMR::CodeGenerator::allocateSpill(int32_t dataSize, bool containsCollectedRefere
 void
 OMR::CodeGenerator::freeSpill(TR_BackingStore *spill, int32_t dataSize, int32_t offset)
    {
-   TR_ASSERT(1 <= dataSize && dataSize <= 16, "assertion failure");
-   TR_ASSERT(offset == 0 || offset == 4, "assertion failure");
-   TR_ASSERT(dataSize + offset <= 16, "assertion failure");
+   TR_ASSERT_FATAL(1 <= dataSize && dataSize <= 64, "assertion failure");
+   TR_ASSERT_FATAL(offset == 0 || offset == 4, "assertion failure");
+   TR_ASSERT_FATAL(dataSize + offset <= 64, "assertion failure");
```
Comment (0xdaryl) [inline compiler/codegen/CodeGenRA.cpp:573]: You may want to take this opportunity to write a more informative assert message for each here.

**Review by 0xdaryl — COMMENTED** [review]
Although it doesn't immediately effect this PR, I'll ask the question now anyway.  Do we align the stack appropriately so that 32- and 64-byte spill temps are naturally aligned on the stack?  There are likely performance implications if not.

**Comment by BradleyWood** [comment]
> Although it doesn't immediately effect this PR, I'll ask the question now anyway. Do we align the stack appropriately so that 32- and 64-byte spill temps are naturally aligned on the stack? There are likely performance implications if not.

I don't think any alignment is happening. However, I believe VRF spills are rare.

**Comment by BradleyWood** [comment]
Jenkins build all

File: compiler/codegen/CodeGenRA.cpp [hunk 3]
```diff
@@ -458,24 +458,37 @@ OMR::CodeGenerator::allocateSpill(int32_t dataSize, bool containsCollectedRefere
...
-       spill = self()->getSpill16FreeList().front();
-       self()->getSpill16FreeList().pop_front();
-     }
+      {
+      if (dataSize <= 4 && !self()->getSpill4FreeList().empty())
+         {
+         spill = self()->getSpill4FreeList().front();
+         self()->getSpill4FreeList().pop_front();
+         }
+      else if (!spill && try8ByteSpills && !self()->getSpill8FreeList().empty())
```
Comment (0xdaryl) [inline compiler/codegen/CodeGenRA.cpp:471]: The `!spill` in all of these `else`s is implied.  Please remove.

Comment (BradleyWood) [reply compiler/codegen/CodeGenRA.cpp]: Got it

**Comment by BradleyWood** [comment]
Jenkins build all

**Review by 0xdaryl — APPROVED** [review]

**Pull request merged**